source/language/openpulse.rst

```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
-means that assigning a frame to an alias (for the purpose of avoiding typing)
-needs to be done with ``const`` not ``let``.
+The ``frame`` type is a virtual resource and the exact precision of these parameters is
+hardware specific. It is thus up to the compiler to choose how to implement the required
+transformations to physical resources in hardware (e.g. mapping multiple frames to a
+single NCO).
 
-.. code-block: none
+Frame Construction
+~~~~~~~~~~~~~~~~~~
 
-   const driveframe = frameof("drive", %0);
+Frames are purely virtual and can be constructed using the ``Frame`` command
 
-Second, frames return references not values. This means that calling the same
-``frameof`` command in different places is the same as assigning the frame to a
-constant and referencing it multiple times.
+.. code-block:: javascript
 
-Frame manipulation
-~~~~~~~~~~~~~~~~~~
+  frame driveframe = Frame(5e9, 0.0);
 
-The (frequency, phase, scale) of a frame can be manipulated throughout program
-by referencing ``.frequency``, ``.phase``, and ``.scale``. Operations must be 
-appropriate for the respective type, ``float`` for frequency, ``angle`` for
-phase, and ``float`` for scale. Again, the exact precision of these calculations
-is hardware specific.
+When instantiated, the frame time starts at 0. ``Frame``s can also be copied using the
+``copy`` command with argument replacement
 
-Here's an example of manipulating the phase to calibrate an ``rz`` gate:
+.. code-block:: javascript
 
-.. code-block:: none
+  frame driveframe1 = Frame(1.0, 5e9, 0.0);
+  frame driveframe2 = copy(driveframe1, phase=driveframe1.phase + pi/2);
 
-   // Shift phase of qubit 0's "drive" frame by pi/4, eg. an rz gate with angle -pi/4
-   frameof("drive", %0).phase += pi/4;
+This will generate a ::math:`pi/2` phase incremented copy of ``driveframe1`` (i.e. with
+the same frequency and time as ``driveframe1``) with the same time as `driveframe1`.
 
-   // Equivalent
-   const drive = frameof("drive", %0);
-   drive.phase += pi/4;
+Like other things in OpenQASM, a ``frame`` is locally scoped unless a ``prelude`` defcal is used
+to instantiate frames in global scope.
 
-   // Define a calibration for the rz gate on all physical qubits
-   defcal rz(angle[20]:theta) %q {
-     frameof("drive", %q).phase -= theta;
+.. code-block:: javascript
+
+   defcal prelude {
+     frame driveframe0 = Frame(1.0, 5e9, 0.0);
+     frame driveframe1 = Frame(1.0, 6e9, 0.0);
+    }
+
+   defcal access_frames $q {
+     frame driveframe2 = copy(driveframe1); // `driveframe1` globally scoped
    }
 
-Here's an example qubit spectroscopy experiment.
 
-.. code-block:: none
+To allow for parameterized access to these globally defined frames, we can also define a ``map``.
+
+.. code-block:: javascript
 
-   qubit q;
+   defcal prelude {
+     frame driveframe0 = Frame(1.0, 5e9, 0.0);
+     frame driveframe1 = Frame(1.0, 6e9, 0.0);
+     ...
+     map[int, frame] frame_mapping = {0: driveframe0, 1: driveframe1};
```

Comment:
Oh, I see. This is another way of approaching this.

Reply:
Possible overextension of the language; the `set_attr` and `get_attr` seems like a legitimate alternative.

Reply:
I like the get/set attr idea vs adding multiple container types to the language. I would also advocate for making it narrower, `getframe` and `setframe` that only work with frames and don't need to work for other types.

Reply:
> getframe and setframe that only work with frames and don't need to work for other types.

This would certainly be easier and be less of a headache. The only other piece that comes to mind would be pulses (eg., applying a scaling to an Xp pulse to get an X90) / virtual channels.